File: fetch_final_odds.js
// 从500.com抓取终盘欧赔，更新到Notion
// 用法: node fetch_final_odds.js

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const NOTION_API_KEY = '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';
const DATABASE_ID = '35491ad7-17ba-81cc-aa04-ce53f7234e17';
const TASKS_DIR = path.join(__dirname, 'tasks');

const headers = {
  'Authorization': 'Bearer ' + NOTION_API_KEY,
  'Notion-Version': '2022-06-28',
  'Content-Type': 'application/json',
};

function notionRequest(method, endpoint, data) {
  return new Promise((resolve, reject) => {
    const body = data ? JSON.stringify(data) : null;
    const req = https.request({
      hostname: 'api.notion.com',
      path: endpoint,
      method,
      headers: {
        ...headers,
        'Content-Length': Buffer.byteLength(body || ''),
      },
    }, res => {
      let d = '';
      res.on('data', c => d += c);
      res.on('end', () => {
        try {
          resolve({ status: res.statusCode, data: JSON.parse(d) });
        } catch (e) {
          resolve({ status: res.statusCode, data: d });
        }
      });
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

// 从500.com抓取欧赔数据
function fetchOddsFrom500(fid) {
  return new Promise((resolve, reject) => {
    const url = `https://odds.500.com/fenxi/ouzhi-${fid}.shtml`;
    const options = {
      hostname: 'odds.500.com',
      path: `/fenxi/ouzhi-${fid}.shtml`,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'zh-CN,zh;q=0.9',
      },
      timeout: 15000,
    };

    const req = https.request(options, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          // 用GBK解码
          const iconv = require('iconv-lite');
          const text = iconv.decode(Buffer.from(data, 'binary'), 'gbk');
          resolve(parseOddsTable(text));
        } catch (e) {
          resolve(null);
        }
      });
    });
    req.on('error', () => resolve(null));
    req.on('timeout', () => { req.destroy(); resolve(null); });
    req.end();
  });
}

// 解析欧赔表格
function parseOddsTable(html) {
  const result = {};
  
  // 匹配公司行: | 1 | 竞彩官方 | x.x | x.x | x.x | x.x | x.x | x.x | ... |
  const companyRegex = /\|\s*(\d+)\s*\|\s*([^|]+)\s*\|([^|]+)\|([^|]+)\|([^|]+)\|([^|]+)\|([^|]+)\|([^|]+)\|/g;
  let match;
  
  while ((match = companyRegex.exec(html)) !== null) {
    const rowNum = parseInt(match[1].trim());
    const name = match[2].trim();
    
    // 提取6个赔率值: 初胜、初平、初负、即胜、即平、即负
    const nums = [];
    for (let i = 3; i <= 8; i++) {
      const val = match[i].trim().replace(/\s/g, '');
      const num = parseFloat(val);
      if (!isNaN(num)) nums.push(num);
    }
    
    if (nums.length >= 6) {
      if (rowNum === 1 || name === '竞彩官方') {
        // 竞彩: 即时胜/平/负 (索引3,4,5)
        result.jc = {
          win: Math.floor(nums[3] * 10) / 10,
          draw: Math.floor(nums[4] * 10) / 10,
          loss: Math.floor(nums[5] * 10) / 10,
        };
      }
      if (rowNum === 6 || name === 'Interwetten') {
        result.iw = {
          win: Math.floor(nums[3] * 10) / 10,
          draw: Math.floor(nums[4] * 10) / 10,
          loss: Math.floor(nums[5] * 10) / 10,
        };
      }
      if (name.includes('平均值') || name.includes('平均')) {
        result.bj = {
          win: Math.floor(nums[3] * 10) / 10,
          draw: Math.floor(nums[4] * 10) / 10,
          loss: Math.floor(nums[5] * 10) / 10,
        };
      }
    }
  }
  
  return result;
}

// 查询Notion记录
async function queryByDate(dateStr) {
  const data = JSON.stringify({
    filter: {
      property: '比赛日期',
      date: { equals: dateStr },
    },
    page_size: 100,
  });
  const res = await notionRequest('POST', '/v1/databases/' + DATABASE_ID + '/query', JSON.parse(data));
  return res.data?.results || [];
}

// 更新Notion记录
async function updateOdds(pageId, odds) {
  const props = {};
  
  if (odds.jc) {
    props['竞彩欧赔胜'] = { number: odds.jc.win };
    props['竞彩欧赔平'] = { number: odds.jc.draw };
    props['竞彩欧赔负'] = { number: odds.jc.loss };
  }
  if (odds.bj) {
    props['百家欧赔胜'] = { number: odds.bj.win };
    props['百家欧赔平'] = { number: odds.bj.draw };
    props['百家欧赔负'] = { number: odds.bj.loss };
  }
  if (odds.iw) {
    props['Interwetten胜'] = { number: odds.iw.win };
    props['Interwetten平'] = { number: odds.iw.draw };
    props['Interwetten负'] = { number: odds.iw.loss };
  }
  
  if (Object.keys(props).length === 0) return false;
  
  const res = await notionRequest('PATCH', `/v1/pages/${pageId}`, { properties: props });
  return res.status === 200;
}

async function main() {
  const dates = ['2026-05-07', '2026-05-08', '2026-05-09', '2026-05-10', '2026-05-11'];
  let totalUpdated = 0;
  let totalFailed = 0;
  
  for (const dateStr of dates) {
    console.log(`\n=== ${dateStr} ===`);
    
    // 查询Notion记录
    const notionRecords = await queryByDate(dateStr);
    console.log(`Notion记录数: ${notionRecords.length}`);
    
    // 读取本地文件获取FID
    const taskDir = path.join(TASKS_DIR, dateStr);
    if (!fs.existsSync(taskDir)) {
      console.log(`  本地目录不存在，跳过`);
      continue;
    }
    
    const reportFiles = fs.readdirSync(taskDir).filter(f => f.endsWith('.md'));
    
    // 建立编号到FID的映射
    const fidMap = {};
    for (const rf of reportFiles) {
      const content = fs.readFileSync(path.join(taskDir, rf), 'utf8');
      const fidMatch = content.match(/FID[::]\s*(\d+)/);
      const numMatch = rf.match(/周[一二三四五六日](\d{3})/);
      if (fidMatch && numMatch) {
        fidMap[numMatch[1]] = fidMatch[1];
      }
    }
    
    // 处理每条记录
    for (const record of notionRecords) {
      const pageId = record.id;
      const notionName = record.properties['Name']?.title?.[0]?.plain_text || '';
      const numMatch = notionName.match(/(\d{3})/);
      const num = numMatch ? numMatch[1] : '';
      
      if (!num || !fidMap[num]) {
        console.log(`  ⏭️ 跳过 ${notionName} (无FID)`);
        continue;
      }
      
      const fid = fidMap[num];
      console.log(`  🔄 ${notionName} (FID:${fid})`);
      
      // 抓取赔率
      const odds = await fetchOddsFrom500(fid);
      
      if (odds && (odds.jc || odds.bj || odds.iw)) {
        const success = await updateOdds(pageId, odds);
        if (success) {
          console.log(`  ✅ 更新成功: 竞彩=${odds.jc?.win}/${odds.jc?.draw}/${odds.jc?.loss} 百家=${odds.bj?.win}/${odds.bj?.draw}/${odds.bj?.loss} IW=${odds.iw?.win}/${odds.iw?.draw}/${odds.iw?.loss}`);
          totalUpdated++;
        } else {
          console.log(`  ❌ 更新失败`);
          totalFailed++;
        }
      } else {
        console.log(`  ⚠️ 未获取到赔率数据`);
        totalFailed++;
      }
      
      // 限速: 每次请求间隔1秒
      await new Promise(r => setTimeout(r, 1000));
    }
  }
  
  console.log(`\n=== 完成 ===`);
  console.log(`成功更新: ${totalUpdated} 条`);
  console.log(`失败: ${totalFailed} 条`);
}

main().catch(console.error);

File: fetch_final_odds.py
# -*- coding: utf-8 -*-
"""从500.com实时获取终盘欧赔，更新到Notion（05-07到05-11）
直接用requests抓取，不调用subprocess
"""
import sys, os, re, json, time
import requests
from bs4 import BeautifulSoup

NOTION_API_KEY = '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'
DATABASE_ID = '35491ad7-17ba-81cc-aa04-ce53f7234e17'
TASKS_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'tasks')

sess = requests.Session()
sess.headers.update({
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'zh-CN,zh;q=0.9',
})

def notion_request(method, endpoint, data=None):
    import http.client
    conn = http.client.HTTPSConnection('api.notion.com')
    headers = {
        'Authorization': f'Bearer {NOTION_API_KEY}',
        'Notion-Version': '2022-06-28',
        'Content-Type': 'application/json',
    }
    body = json.dumps(data).encode('utf-8') if data else None
    conn.request(method, endpoint, body, headers)
    res = conn.getresponse()
    d = res.read().decode('utf-8')
    try:
        return {'status': res.status, 'data': json.loads(d)}
    except:
        return {'status': res.status, 'data': d}

def query_by_date(date_str):
    data = {
        'filter': {'property': '比赛日期', 'date': {'equals': date_str}},
        'page_size': 100,
    }
    res = notion_request('POST', f'/v1/databases/{DATABASE_ID}/query', data)
    return res.get('data', {}).get('results', [])

def fetch_odds(fid):
    """从500.com抓取欧赔数据（同step146_extractor.py逻辑）"""
    try:
        r = sess.get(f'https://odds.500.com/fenxi/ouzhi-{fid}.shtml', timeout=15)
        r.encoding = 'gbk'
        soup = BeautifulSoup(r.text, 'html.parser')
        
        all_companies = []
        for table in soup.find_all('table'):
            for tr in table.find_all('tr'):
                tds = tr.find_all('td')
                if len(tds) < 12:
                    continue
                td0 = tds[0].get_text().strip()
                td1 = tds[1].get_text().strip()
                nums = []
                for idx in [3, 4, 5, 6, 7, 8]:
                    val = tds[idx].get_text().strip().replace('\xa0', '')
                    try:
                        nums.append(float(val))
                    except:
                        pass
                if len(nums) >= 6:
                    if td0.isdigit():
                        all_companies.append({'row_num': int(td0), 'name': td1,
                            'lw': nums[3], 'ld': nums[4], 'll': nums[5]})
                    elif '平均值' in td1:
                        all_companies.append({'row_num': 99, 'name': '百家平均',
                            'lw': nums[3], 'ld': nums[4], 'll': nums[5]})
        
        odds = {}
        for c in all_companies:
            if c['row_num'] == 1:
                odds['jc'] = {
                    'win': round(c['lw'], 1),
                    'draw': round(c['ld'], 1),
                    'loss': round(c['ll'], 1),
                }
            elif c['row_num'] == 6:
                odds['iw'] = {
                    'win': round(c['lw'], 1),
                    'draw': round(c['ld'], 1),
                    'loss': round(c['ll'], 1),
                }
            elif c['row_num'] == 99:
                odds['bj'] = {
                    'win': round(c['lw'], 1),
                    'draw': round(c['ld'], 1),
                    'loss': round(c['ll'], 1),
                }
        
        return odds
    except Exception as e:
        return {}

def update_odds(page_id, odds):
    props = {}
    if odds.get('jc'):
        props['终盘竞彩欧赔胜'] = {'number': odds['jc']['win']}
        props['终盘竞彩欧赔平'] = {'number': odds['jc']['draw']}
        props['终盘竞彩欧赔负'] = {'number': odds['jc']['loss']}
    if odds.get('bj'):
        props['终盘百家欧赔胜'] = {'number': odds['bj']['win']}
        props['终盘百家欧赔平'] = {'number': odds['bj']['draw']}
        props['终盘百家欧赔负'] = {'number': odds['bj']['loss']}
    if odds.get('iw'):
        props['终盘Interwetten胜'] = {'number': odds['iw']['win']}
        props['终盘Interwetten平'] = {'number': odds['iw']['draw']}
        props['终盘Interwetten负'] = {'number': odds['iw']['loss']}
    
    if not props:
        return False
    
    res = notion_request('PATCH', f'/v1/pages/{page_id}', {'properties': props})
    return res['status'] == 200

def main():
    dates = ['2026-05-07', '2026-05-08', '2026-05-09', '2026-05-10', '2026-05-11']
    total_updated = 0
    total_failed = 0
    
    for date_str in dates:
        print(f'\n=== {date_str} ===')
        notion_records = query_by_date(date_str)
        print(f'Notion记录数: {len(notion_records)}')
        
        task_dir = os.path.join(TASKS_DIR, date_str)
        if not os.path.exists(task_dir):
            print(f'  本地目录不存在，跳过')
            continue
        
        report_files = [f for f in os.listdir(task_dir) if f.endswith('.md')]
        fid_map = {}
        for rf in report_files:
            content = open(os.path.join(task_dir, rf), 'r', encoding='utf-8').read()
            fid_match = re.search(r'FID[::]\s*(\d+)', content)
            num_match = re.search(r'周[一二三四五六日](\d{3})', rf)
            if fid_match and num_match:
                fid_map[num_match.group(1)] = fid_match.group(1)
        
        for record in notion_records:
            page_id = record['id']
            notion_name = record.get('properties', {}).get('Name', {}).get('title', [{}])[0].get('plain_text', '')
            num_match = re.search(r'(\d{3})', notion_name)
            num = num_match.group(1) if num_match else ''
            
            if not num or num not in fid_map:
                print(f'  [SKIP] {notion_name} (无FID)')
                continue
            
            fid = fid_map[num]
            odds = fetch_odds(fid)
            
            if odds:
                success = update_odds(page_id, odds)
                if success:
                    jc = odds.get('jc', {})
                    bj = odds.get('bj', {})
                    iw = odds.get('iw', {})
                    jc_s = f"{jc.get('win')}/{jc.get('draw')}/{jc.get('loss')}" if jc else '-'
                    bj_s = f"{bj.get('win')}/{bj.get('draw')}/{bj.get('loss')}" if bj else '-'
                    iw_s = f"{iw.get('win')}/{iw.get('draw')}/{iw.get('loss')}" if iw else '-'
                    print(f'  [OK] {notion_name} 竞彩:{jc_s} 百家:{bj_s} IW:{iw_s}')
                    total_updated += 1
                else:
                    print(f'  [FAIL] {notion_name}')
                    total_failed += 1
            else:
                print(f'  [SKIP] {notion_name} 未获取到')
                total_failed += 1
            
            time.sleep(0.5)
    
    print(f'\n=== 完成 ===')
    print(f'成功: {total_updated} 失败: {total_failed}')

if __name__ == '__main__':
    main()
